### ISSUE TITLE
Negative duration returned for streaming mp3 on iOS device

### ISSUE THREAD

**Issue by suragch** [ryanheise/just_audio#68]
I'm streaming an mp3 file from my server. Using just_audio works as expected on an Android device, but when I try to get the audio duration on an iOS device (simulator), a negative value is returned.

This does not happen when using the [example](https://pub.dev/packages/just_audio#-example-tab-) audio file `"https://s3.amazonaws.com/scifri-episodes/scifri20181123-episode.mp3"`, so I'm not sure if this is a bug in the plugin or a bug in how my server is returning the headers. 

If it is a bug in my server setup, what are the required headers that iOS needs to obtain the mp3 audio duration?

**Comment by ryanheise** [comment]
Could you provide a link to an example mp3 that illustrates the problem?

**Comment by suragch** [comment]
Here is an example:

```
_totalDuration = await _player.setUrl('https://yorool.com/api/v1/audio/book/example.mp3');
print('total duration: $_totalDuration');
```

This gives the following duration:

```
total duration: -596:31:23.648000
```

The audio still plays but I can't manage the seek bar since I don't know the total duration.

**Comment by rohansohonee** [comment]
@suragch

Please use a real iOS device for testing. Simulators/Emulators often don't work as expected.
I am just sharing my experience.

**Comment by ryanheise** [comment]
@suragch , another thing to confirm is, when you say you are "streaming" the mp3, how are you streaming it? Do you mean you're not just offering the mp3 at a URL, but you're using some sort of streaming protocol? If so, which one?

**Comment by suragch** [comment]
@rohansohonee Good point. The simulator can behave differently than a real device, though it is interesting that the android emulator works while the iOS simulator doesn't. I tested it from Xcode on an iPhone 4s (the only iOS device I have) and got a cryptic error message from Xcode:

```
2020-04-07 13:02:34.302 Runner[201:3943] time discontinuity detected: -656
```

Not even the audio played this time.

**Comment by suragch** [comment]
@ryanheise I'm a newcomer to streaming. I assumed that streaming just meant setting the URL and then calling `player.play()` rather than downloading first before calling play. I'm not using any sort of streaming protocol client side other than what just_audio provides.

**Comment by simdev0** [comment]
I'm also having this issue. I use .setUrl(url).then((duration) { /* set duration here */ } and found that it comes up with a negative value.

**Comment by ryanheise** [comment]
> Here is an example:
> 
> ```
> _totalDuration = await _player.setUrl('https://yorool.com/api/v1/audio/book/example.mp3');
> print('total duration: $_totalDuration');
> ```
> 
> This gives the following duration:
> 
> ```
> total duration: -596:31:23.648000
> ```
> 
> The audio still plays but I can't manage the seek bar since I don't know the total duration.

I tested on the iOS 10.1 / iPhone 7 plus simulator, but it worked:

```
total duration: 0:02:56.613000
```

@simdev0 do you also have a URL? Does it give the exact same negative number or a different one?

And just to confirm, you did add the following to your `Info.plist`?

```xml
<key>NSAppTransportSecurity</key>
<dict>
    <key>NSAllowsArbitraryLoads</key>
    <true/>
    <key>NSAllowsArbitraryLoadsForMedia</key>
    <true/>
</dict>
```

**Comment by suragch** [comment]
@ryanheise Sorry, the link I provided is no longer valid because I changed the server configuration a couple days ago. Before the mp3 didn't have Content-Length header but now it does.

To reproduce the issue, any response without a Content-Length header value should do the same thing. 

I wrote more about the server side solution over here: https://stackoverflow.com/questions/61318517/files-served-with-aqueduct-dont-have-a-content-length-header

I'm not sure if it is possible to stream a file without a Content-Length header on iOS. Not everyone can change the server, so if there is some way to handle it in the plugin that would be nice.

**Comment by ryanheise** [comment]
Probably the percentage of misconfigured servers is very very low because it would have to be a custom built server (nginx and apache would handle it correctly by default), such as for APIs, and there's a high chance that frontend apps that consume those APIs will be written by the same developer who has some control over both ends.

In the rare case where someone has written an API intended to be consumed by 3rd parties, and the API developer has not included this expected header, and the developer is not responding to support emails to fix it, I suppose one alternative for the 3rd party is to write some code to test if the server didn't provide the expected header and if not, just download the file locally and then play it locally.

Thoughts?

**Comment by suragch** [comment]
From the plugin side, it might be useful to return null for the Duration (rather than a negative number) if the content length is unknown (and describe that behavior in the documentation).

**Comment by ryanheise** [comment]
I agree, I'll have a go at it next time I'm on MacInCloud.

Although it will still be useful to have an example URL to reproduce the issue.

**Comment by suragch** [comment]
I tried finding another source but wasn't able to just now. I can't easily change my server back right now without messing up the client apps. Hopefully @simdev0 can provide a link.

**Comment by simdev0** [comment]
@ryanheise Here's the link I was using: https://smasby-api20191025092629.azurewebsites.net/api/talkFile/Bible%20Study/Ancient%20Flaws,%20Modern%20Solutions/meriam's%20bible%20study%20(abraham).mp3 
Something to note, I got the exact same negative value as when trying to get the duration of that URL. 

@suragch just a question, you said you retrieved that negative value when you had a response without a content-length header. When using the link I provided in the browser, I checked the network and saw that it contains the content-length header in the response, but I'm still getting the same issue. Is it something I need to change in the code perhaps?

**Comment by suragch** [comment]
@simdev0, I don't know. I wouldn't have expected a negative value for that. However, because of the problems I was having with streaming, I ended up just downloading the file and then playing it from local storage with the audioplayers plugin. 
https://medium.com/flutter-community/how-to-show-download-progress-in-a-flutter-app-8810e294acbd

**Comment by ryanheise** [comment]
This should be implemented correctly in the 0.2.0 release, where `null` is returned for an unknown duration. I'll close this, but please let me know if the issue persists and I'll reopen.
